### PR TITLE
feat: Improve Ratings on desktop (fix #2564)

### LINKED ISSUE #2564: Implement a desktop "all reviews" page
Implement the review page for desktop, relates to #2552.

### PR DESCRIPTION
**Pull request by tofumatt**
Works toward #2564, but the ratings grouped by star number is a lot more work to implement, and this is already a fair bit big.

So, this leaves out the ratings grouped by stars. But this helps pull the add-on detail header together and is most of this issue. I'm gonna do the grouping separately because it requires a bit of saga/redux/etc. work.

### Review info on details
screenshot 2017-10-24 12 07 20

### All reviews page
screenshot 2017-10-24 12 07 28

### Mobile
screen shot 2017-10-24 at 12 07 43

### CODE REVIEW

File: src/amo/components/AddonReviewList/styles.scss [hunk 1]
```diff
@@ -1,57 +1,63 @@
...
-      word-wrap: break-word;
-    }
+  padding: 12px 0;
+}
+
+.AddonReviewList-header-icon {
+  @include margin-end(18px);
+}
+
+.AddonReviewList-header-addonName {
```
Comment (willdurand) [inline src/amo/components/AddonReviewList/styles.scss]: I dont know how to do it but ideally, the text next to the icon should be vertically centered.

screen shot 2017-10-24 at 15 03 31

**Comment by willdurand** [comment]
Medium size also compress the rating widget:

screen shot 2017-10-24 at 15 07 46

File: src/amo/components/AddonMeta/styles.scss [hunk 2]
```diff
@@ -0,0 +1,45 @@
+@import "~core/css/inc/mixins";
+@import "~amo/css/inc/vars";
+@import "~ui/css/vars";
+
+$background: #f8f9fa;
+
+.AddonMeta .Card-contents {
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/amo/components/AddonMeta/styles.scss]: It would and I tried but the content and layout is pretty different (right now). I wanted to land this so I admittedly rushed. I'd rather do that as a follow-up, if at all possible.

File: src/amo/components/AddonMeta/index.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    /* eslint-disable react/no-danger */
+    return (
+      <Card className="AddonMeta">
+        <div className="AddonMeta-item AddonMeta-users">
+          <h3 className="visually-hidden">{i18n.gettext('Used by')}</h3>
+          {userCount ? (
+            <div
+              className="AddonMeta-text AddonMeta-user-count"
+              dangerouslySetInnerHTML={sanitizeHTML(userCount, ['span'])}
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/amo/components/AddonMeta/index.js]: Hmm, that would make sense. This component was a bit messy looking back at it and I guess I was trying to fit things into how it already worked. I'll give your way a good and maybe we can even re-use the component after all.

File: src/amo/components/AddonReviewList/index.js [hunk 4]
```diff
@@ -167,17 +207,33 @@ export class AddonReviewListBase extends React.Component<Props> {
...
+              {authorsHTML}
+            </div>
           </div>
-        </div>
-        <CardList>
+
+          {addon ? (
+            <RatingManager
+              addon={addon}
+              location={{}}
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/amo/components/AddonReviewList/index.js]: It should be better than this (sorry, I forgot about this). The `RatingManager` sends location to the `AuthenticateButton`. It means after signing in you get sent to the old URL you were on. I'd like to fix it eventually (the Authenticate button should really get the router location itself), but didn't want to overload the patch.

File: src/amo/components/AddonReviewList/styles.scss [hunk 5]
```diff
@@ -1,57 +1,63 @@
-@import "~amo/css/inc/vars";
 @import "~core/css/inc/mixins";
+@import "~amo/css/inc/vars";
+@import "~ui/css/vars";
 
 .AddonReviewList {
-  padding: $padding-page;
+  margin: $padding-page;
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/amo/components/AddonReviewList/styles.scss]: It's a top-level element and margin seems to make more sense but I guess I'll restore it. Looks the same.

**Comment by tofumatt** [comment]
Thanks for the review. I'm so tired from immigration nonsense I rushed this patch. I'll do my best to get it ready for another r? ASAP.

File: src/amo/components/AddonReviewList/index.js [hunk 6]
```diff
@@ -150,12 +149,53 @@ export class AddonReviewListBase extends React.Component<Props> {
...
+    const authorProps = {};
+    if (addon && addon.authors) {
+      const authorList = addon.authors.map(
+        (author) => oneLine`
+        <a
+          class="AddonReviewList-addon-author-link"
+          href="${author.url}"
+        >
+          ${author.name}
+        </a>`
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/amo/components/AddonReviewList/index.js]: No worries, fixed here.

File: src/amo/components/AddonReviewList/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/amo/components/AddonReviewList/index.js]: Oh, but actually we already have it as a prop so it's fine.

File: src/amo/components/AddonReviewList/styles.scss (hunk 1 above)
Comment (tofumatt) [reply src/amo/components/AddonReviewList/styles.scss]: Frustratingly the mocks have much larger icons than I think we actually have, so they'll look pretty blurry. I'll give that a go though.

**Comment by tofumatt** [comment]
Sorry this took forever; I had to rebase, deal with tests, and also I had a social thing I had to get to in the evening. All ready for another r? now, thanks.

**Comment by tofumatt** [comment]
Updated screenshots, sorry, I forgot:
screen shot 2017-10-25 at 01 23 59
screen shot 2017-10-25 at 01 23 59

File: src/amo/components/AddonReviewList/index.js [hunk 7]
```diff
@@ -150,12 +149,52 @@ export class AddonReviewListBase extends React.Component<Props> {
...
+      );
     } else {
       addonName = <LoadingText />;
+      reviewCountHTML = <LoadingText />;
+    }
+
+    const authorProps = {};
+    if (addon && addon.authors) {
+      const authorList = addon.authors.map((author) => {
+        return author.url ? oneLine`
```
Comment (willdurand) [inline src/amo/components/AddonReviewList/index.js]: This is not really readable. Any chance to have a `if`?

File: src/amo/components/AddonReviewList/styles.scss (hunk 1 above)
Comment (willdurand) [reply src/amo/components/AddonReviewList/styles.scss]: Looks like this is not solved, right?

File: src/ui/components/MetadataCard/index.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+/* @flow */
+import React from 'react';
+
+import LoadingText from 'ui/components/LoadingText';
+
+import './styles.scss';
+
+
+type Props = {|
+  metadata: Array<Object>,
```
Comment (willdurand) [inline src/ui/components/MetadataCard/index.js]: `Object` is actually `{| title: string, content: any |}`?

File: src/ui/components/MetadataCard/index.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+|};
+
+const MetadataCard = (props: Props) => {
+  if (!props.metadata) {
+    throw new Error('The metadata property is required');
+  }
+
+  return (
+    <div className="MetadataCard">
+      {props.metadata.map(({ content, title } = {}) => {
```
Comment (willdurand) [inline src/ui/components/MetadataCard/index.js]: This is weird no? I think each `metadata` entry should always contain these properties. I don't see where it is missing in the current code.

File: src/ui/components/MetadataCard/styles.scss [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+@import "~amo/css/inc/vars";
+@import "~ui/css/vars";
+
+.MetadataCard {
+  background-color: $grey-10;
+  border-radius: $border-radius-m;
+  display: flex;
+  flex-direction: row;
+  justify-content: space-between;
+  margin-bottom: 10px;
```
Comment (willdurand) [inline src/ui/components/MetadataCard/styles.scss]: Why is this needed? It looks not aligned because of it on small screens:

screen shot 2017-10-25 at 15 36 50

File: tests/unit/amo/components/TestCollection.js [hunk 11]
```diff
@@ -66,8 +67,18 @@ describe(__filename, () => {
 
     // We display 5 items on the detail card: title, description, number of
     // addons, creator, and last modified date.
+    // 3 items are rendered by `MetadataCard` though and will render
+    // as `LoadingText` if null; so we just make sure it has `null`
+    // props for the last three.
```
Comment (willdurand) [inline tests/unit/amo/components/TestCollection.js]: :+1:

File: src/ui/components/MetadataCard/index.js (hunk 9 above)
Comment (willdurand) [reply src/ui/components/MetadataCard/index.js]: Also I dont find any test cases for that.

**Comment by willdurand** [comment]
Screenshots for the issue I mentioned:

screen shot 2017-10-25 at 15 45 10
screen shot 2017-10-25 at 15 45 15

File: src/ui/components/MetadataCard/styles.scss (hunk 10 above)
Comment (tofumatt) [reply src/ui/components/MetadataCard/styles.scss]: I think this was just copied from the old styles but I'll check into what it's doing.

Comment (tofumatt) [reply src/ui/components/MetadataCard/styles.scss]: Turns out it was not needed, thanks.

File: src/ui/components/MetadataCard/index.js (hunk 9 above)
Comment (tofumatt) [reply src/ui/components/MetadataCard/index.js]: Agreed, added some tests for that.

**Comment by tofumatt** [comment]
I've made some tweaks to the alignment and icon:

screenshot 2017-10-26 12 26 29

Fixed the margins too:

screen shot 2017-10-26 at 12 28 42

**Comment by willdurand** [comment]
Looks tip top 👍
